src/Http/OpenApi/src/OpenApiGenerator.cs

```diff
@@ -0,0 +1,497 @@
+            };
+        }
+
+        if (requestBodyParameter is not null)
+        {
+            if (requestBodyContent.Count == 0)
+            {
+                var isFormType = requestBodyParameter.ParameterType == typeof(IFormFile) || requestBodyParameter.ParameterType == typeof(IFormFileCollection);
+                var hasFormAttribute = requestBodyParameter.GetCustomAttributes().OfType<IFromFormMetadata>().FirstOrDefault() != null;
+                if (isFormType || hasFormAttribute)
+                {
+                    requestBodyContent["multipart/form-data"] = new OpenApiMediaType
+                    {
+                        Schema = new OpenApiSchema
+                        {
+                            Type = requestBodyParameter.ParameterType.Name
+                        }
+                    };
+                }
+                else
+                {
+                    requestBodyContent["application/json"] = new OpenApiMediaType
+                    {
+                        Schema = new OpenApiSchema
+                        {
+                            Type = requestBodyParameter.ParameterType.Name
+                        }
+                    }; ;
+                }
+            }
+
+            var nullabilityContext = new NullabilityInfoContext();
+            var nullability = nullabilityContext.Create(requestBodyParameter);
+            var allowEmpty = requestBodyParameter.GetCustomAttributes().OfType<IFromBodyMetadata>().SingleOrDefault()?.AllowEmpty ?? false;
+            var isOptional = requestBodyParameter.HasDefaultValue
+                || nullability.ReadState != NullabilityState.NotNull
+                || allowEmpty;
+
+            return new OpenApiRequestBody
+            {
+                Required = !isOptional,
+                Content = requestBodyContent
+            };
+        }
+
+        return null;
+    }
+
+    private IList<OpenApiTag> GetOperationTags(MethodInfo methodInfo, EndpointMetadataCollection metadata)
+    {
+        var tags = metadata.GetMetadata<ITagsMetadata>();
+        string controllerName;
+
+        if (methodInfo.DeclaringType is not null && !TypeHelper.IsCompilerGeneratedType(methodInfo.DeclaringType))
+        {
+            controllerName = methodInfo.DeclaringType.Name;
+        }
+        else
+        {
+            // If the declaring type is null or compiler-generated (e.g. lambdas),
+            // group the methods under the application name.
+            controllerName = _environment?.ApplicationName ?? string.Empty;
+        }
+
+        return tags is not null
+            ? tags.Tags.Select(tag => new OpenApiTag() { Name = tag }).ToList()
+            : new List<OpenApiTag>() { new OpenApiTag() { Name = controllerName } };
+    }
+
+    private static OperationType MapHttpMethodToOperationType(string httpMethod)
+    {
+        return httpMethod switch
+        {
+            "GET" => OperationType.Get,
```

Comment:
nit: Don't we have constants or `static readonly` properties somewhere for all of these `string`s❔